urllib3/sessionmanager.py

```diff
@@ -0,0 +1,78 @@
+from .util.retry import Retry, handle_redirect
+from .request import RequestMethods, Request
+from .poolmanager import PoolManager, ProxyManager
+from .util.context import Context
+
+
+class SessionManager(RequestMethods):
+    """
+    Allows arbitrary requests while maintaining browser-session-like
+    context across those requests.
+
+    :param headers:
+        Headers to include with all requests, unless other headers are
+        given explcitly
+
+    :param cookie_jar:
+        A predefined :class:`CookieJar` object to use in the session;
+        if not provided, a new one will be created.
+
+    :param \**pm_kwargs:
+        Additional paramaters to be passed to the encapsulated
+        :class:`.PoolManager` or :class:`.ProxyManager` instance.
+    """
+    def __init__(self, headers=None, cookie_jar=None, **pm_kwargs):
+
+        super(SessionManager, self).__init__(headers=headers)
+
+        # Detect if we should create a PoolManager or a ProxyManager
+        if pm_kwargs.get('proxy_url', None) is None:
```

Comment:
Makes sense! Any thoughts on the classmethod verbiage? I was thinking something along the lines of `SessionManager.with_pool_manager(*args, **kwargs)`.

Alternatively, would you be open to optionally taking an existing *Manager, and if none was received, initialize a standard `PoolManager` with the received args? That way, we'd get one-step init for the common use case, and would also have a canonical way to initialize with an alternative of any appropriate type. It could look something like this:

    class SessionManager(object):

        def __init__(self, manager=None, headers=None, cookie_jar=None, manager_class=PoolManager, **manager_kw):
            if manager:
                self.manager = manager
            else:
                self.manager = manager_class(**manager_kw)
            # Do other stuff

Something like that could be subclassed relatively simply to use any variety of manager:

    class ProxySessionManager(SessionManager):

        def __init__(self, **kw):
            kw['manager_class'] = ProxyManager
            super(ProxySessionManager, self).__init__(**kw)